rdp-server/eventloop.cpp

```diff
@@ -78,112 +83,67 @@ struct _ogon_event_source {
 };
 
 ogon_event_loop *eventloop_create(void) {
-	ogon_event_loop *ret = calloc(1, sizeof(ogon_event_loop) );
+	auto ret = new (ogon_event_loop);
 	if (!ret) {
 		WLog_ERR(TAG, "error creating event loop");
-		return NULL;
+		return nullptr;
 	}
 
 #ifdef HAVE_EPOLL_H
 	ret->epollfd = epoll_create(1);
 	if (ret->epollfd < 0) {
-		WLog_ERR(TAG, "error creating epollfd, epoll_create returned %d", errno);
-		goto out_free_loop;
+		WLog_ERR(
+				TAG, "error creating epollfd, epoll_create returned %d", errno);
+		goto fail;
 	}
 #else
 	FD_ZERO(&ret->readset);
 	FD_ZERO(&ret->writeset);
 	FD_ZERO(&ret->exceptset);
 #endif
 
-	ret->sources = LinkedList_New();
-	if (!ret->sources) {
-		WLog_ERR(TAG, "error creating sources LinkedList");
-		goto out_epoll;
-	}
-
-	ret->cleanups = LinkedList_New();
-	if (!ret->cleanups) {
-		WLog_ERR(TAG, "error creating cleanups LinkedList");
-		goto out_sources;
-	}
-
-	ret->rescheduled = LinkedList_New();
-	if (!ret->rescheduled) {
-		WLog_ERR(TAG, "error creating rescheduled LinkedList");
-		goto out_cleanups;
-	}
-
 	return ret;
 
-out_cleanups:
-	LinkedList_Free(ret->cleanups);
-out_sources:
-	LinkedList_Free(ret->sources);
-out_epoll:
-#ifdef HAVE_EPOLL_H
-	close(ret->epollfd);
-out_free_loop:
-#endif
-	free(ret);
-	return NULL;
+fail:
+	eventloop_destroy(&ret);
+	return nullptr;
 }
 
 static void treat_cleanups(ogon_event_loop *evloop) {
-	ogon_event_source *source;
-
-	LinkedList_Enumerator_Reset(evloop->cleanups);
-	while(LinkedList_Enumerator_MoveNext(evloop->cleanups)) {
-		source = LinkedList_Enumerator_Current(evloop->cleanups);
+	for (auto source : evloop->cleanups) {
 		if (source->dtor) {
 			source->dtor(source);
 		}
-		LinkedList_Remove(evloop->sources, source);
-
-		free(source);
+		evloop->sources.erase(source);
 	}
-
-	LinkedList_Clear(evloop->cleanups);
+	evloop->cleanups.clear();
 }
 
 static void treat_rescheduled(ogon_event_loop *evloop) {
-	ogon_event_source *source;
-
-	while (LinkedList_Count(evloop->rescheduled) > 0)	{
-		source = LinkedList_First(evloop->rescheduled);
-
+	for (auto source : evloop->rescheduled) {
 		if (!source->markedForRemove) {
 			int mask = source->rescheduleMask;
 			source->rescheduleMask = 0;
 
 			source->callback(mask, source->fd, source->handle, source->data);
 		}
-
-		LinkedList_RemoveFirst(evloop->rescheduled);
 	}
+	evloop->rescheduled.clear();
 }
 
 void eventloop_destroy(ogon_event_loop **pevloop) {
 	ogon_event_loop *evloop = *pevloop;
 
-	if (!evloop)
-		return;
+	if (!evloop) return;
 
 #ifdef HAVE_EPOLL_H
 	close(evloop->epollfd);
 #endif
 
 	treat_cleanups(evloop);
-	LinkedList_Free(evloop->cleanups);
 
-	LinkedList_Enumerator_Reset(evloop->sources);
-	while(LinkedList_Enumerator_MoveNext(evloop->sources)) 	{
-		free(LinkedList_Enumerator_Current(evloop->sources));
-	}
-	LinkedList_Free(evloop->sources);
-	LinkedList_Free(evloop->rescheduled);
-	free(evloop);
-	*pevloop = NULL;
+	delete (evloop);
```

Comment:
if I've not missed something, the sources contained in evloop->sources are leaked